# 2.2 Complex Machine Learning Models and Keras Part 1

#### 1. Importing Libraries and Datas)
#### 2. Data Wrangling
#### 3. Reshaping for modeling
#### 4. Data Split
#### 5. Creating Keras Model
#### 6. Compiling and Running
#### 7. Creating Confusion Matrix

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import os
import operator
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow import keras
from numpy import unique
from numpy import reshape
from keras.models import Sequential
from keras.layers import Conv1D, Conv2D, Dense, BatchNormalization, Flatten, MaxPooling1D, Dropout
from keras.utils import to_categorical
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

In [4]:
# Ensures the charts created are displayed in the notebook without the need to "call" them.
%matplotlib inline

# Turning off warning feature
import warnings
warnings.filterwarnings('ignore')

# Supress scientific notation for easier analysis profiling
pd.set_option('display.float_format', '{:.2f}'.format)

pd.set_option('display.max_columns', None)

In [5]:
#Create a path to where your data is stored.
path = r'C:\Users\IVON6\ClimateWins'

In [6]:
#Read in the European weather data and Answers Data set.

climate = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'Dataset-weather-prediction-dataset-processed.csv'))
answers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'Weather_Prediction_Pleasant_Weather.csv'))

In [7]:
climate.head()

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0  19600101      1                  7              2.10            0.85   
1  19600102      1                  6              2.10            0.84   
2  19600103      1                  8              2.10            0.90   
3  19600104      1                  3              2.10            0.92   
4  19600105      1                  6              2.10            0.95   

   BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0            1.02                    0.32                 0.09   
1            1.02                    0.36                 1.05   
2            1.02                    0.18                 0.30   
3            1.02                    0.58                 0.00   
4            1.02                    0.65                 0.14   

   BASEL_snow_depth  BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  \
0                 0            0.70             6.50            0.80   
1                 0            1.10             6.10            3.30   
2                 0            0.00             8.50            5.10   
3                 0            4.10             6.30            3.80   
4                 0            5.40             3.00           -0.70   

   BASEL_temp_max  BELGRADE_cloud_cover  BELGRADE_humidity  BELGRADE_pressure  \
0           10.90                     1               0.81               1.02   
1           10.10                     6               0.84               1.02   
2            9.90                     6               0.77               1.02   
3           10.60                     8               0.93               1.03   
4            6.00                     8               0.99               1.03   

   BELGRADE_global_radiation  BELGRADE_precipitation  BELGRADE_sunshine  \
0                       0.88                    0.00               7.00   
1                       0.25                    0.00               0.00   
2                       0.67                    0.00               3.50   
3                       0.25                    0.00               0.00   
4                       0.25                    0.06               0.00   

   BELGRADE_temp_mean  BELGRADE_temp_min  BELGRADE_temp_max  \
0                3.70              -0.90               7.90   
1                2.90               2.20               4.40   
2                3.10              -0.50               6.40   
3                2.00              -2.00               3.00   
4                2.00               0.70               2.80   

   BUDAPEST_cloud_cover  BUDAPEST_humidity  BUDAPEST_pressure  \
0                     4               0.67               1.02   
1                     4               0.67               1.02   
2                     4               0.67               1.02   
3                     4               0.67               1.02   
4                     4               0.67               1.02   

   BUDAPEST_global_radiation  BUDAPEST_precipitation  BUDAPEST_sunshine  \
0                       0.44                    0.01               2.30   
1                       0.18                    0.31               0.00   
2                       0.30                    0.00               0.60   
3                       0.19                    0.00               0.00   
4                       0.19                    0.00               0.00   

   BUDAPEST_temp_mean  BUDAPEST_temp_min  BUDAPEST_temp_max  \
0                2.40              -0.40               5.10   
1                2.30               1.40               3.10   
2                2.70               1.70               5.30   
3                2.00               0.40               4.40   
4                2.50               1.10               5.30   

   DEBILT_cloud_cover  DEBILT_wind_speed  DEBILT_humidity  DEBILT_pressure  \
0                   7               7.70             0.85             1.00   
1                   8           

In [8]:
answers.head()

DATE  BASEL_pleasant_weather  BELGRADE_pleasant_weather  \
0  19600101                       0                          0   
1  19600102                       0                          0   
2  19600103                       0                          0   
3  19600104                       0                          0   
4  19600105                       0                          0   

   BUDAPEST_pleasant_weather  DEBILT_pleasant_weather  \
0                          0                        0   
1                          0                        0   
2                          0                        0   
3                          0                        0   
4                          0                        0   

   DUSSELDORF_pleasant_weather  HEATHROW_pleasant_weather  \
0                            0                          0   
1                            0                          0   
2                            0                          0   
3                            0                          0   
4                            0                          0   

   KASSEL_pleasant_weather  LJUBLJANA_pleasant_weather  \
0                        0                           0   
1                        0                           0   
2                        0                           0   
3                        0                           0   
4                        0                           0   

   MAASTRICHT_pleasant_weather  MADRID_pleasant_weather  \
0                            0                        0   
1                            0                        0   
2                            0                        0   
3                            0                        0   
4                            0                        0   

   MUNCHENB_pleasant_weather  OSLO_pleasant_weather  \
0                          0                      0   
1                          0                      0   
2                          0                      0   
3                          0                      0   
4                          0                      0   

   SONNBLICK_pleasant_weather  STOCKHOLM_pleasant_weather  \
0                           0                           0   
1                           0                           0   
2                           0                           0   
3                           0                           0   
4                           0                           0   

   VALENTIA_pleasant_weather  
0                          0  
1                          0  
2                          0  
3                          0  
4                          0

In [9]:
climate.shape

(22950, 170)

In [10]:
answers.shape

(22950, 16)

## 2. Data Wrangling
Ensure data is structured with correct shape to feed the deep learning model:
- Drop 3 weather stations not included in answers.
- Remove 2 types of observations (columns) missing multiple entries for most stations.
- Fill in 3 individual observations assuming nearby stations have similar weather.
- Drop DATE and MONTH from observations and DATE from predictions data set.
- Export dataset as "Cleaned" version. X shape should be (22950, 135) and y shape should be (22950, 15).

In [12]:
# Drop 3 weather stations not included in answers. # These are unique columns in answers dataframe
awn_col = []
for i in answers.columns.tolist():
    awn_col.append(i.split('_')[0])

awn_col = set(awn_col)
awn_col

{'BASEL',
 'BELGRADE',
 'BUDAPEST',
 'DATE',
 'DEBILT',
 'DUSSELDORF',
 'HEATHROW',
 'KASSEL',
 'LJUBLJANA',
 'MAASTRICHT',
 'MADRID',
 'MUNCHENB',
 'OSLO',
 'SONNBLICK',
 'STOCKHOLM',
 'VALENTIA'}

In [13]:
# Finding all stations in Climate DataFrame
col = climate.columns.tolist()
cli_col = [i.split('_')[0] for i in col]
cli_col = set(cli_col)

In [14]:
# Finding columns that are not in both DataFrames
dif = []
for i in cli_col:
    if i not in awn_col:
        dif.append(i)
dif

['TOURS', 'GDANSK', 'MONTH', 'ROMA']

In [15]:
# Searching for stations that are not Awnsers DataFrame

import re

del_colum = []
for i in col:
    if re.search('GDANSK|ROMA|TOURS|MONTH', str(i)):
        del_colum.append(i)
del_colum.append('DATE')
del_colum

['MONTH',
 'GDANSK_cloud_cover',
 'GDANSK_humidity',
 'GDANSK_precipitation',
 'GDANSK_snow_depth',
 'GDANSK_temp_mean',
 'GDANSK_temp_min',
 'GDANSK_temp_max',
 'ROMA_cloud_cover',
 'ROMA_wind_speed',
 'ROMA_humidity',
 'ROMA_pressure',
 'ROMA_sunshine',
 'ROMA_temp_mean',
 'TOURS_wind_speed',
 'TOURS_humidity',
 'TOURS_pressure',
 'TOURS_global_radiation',
 'TOURS_precipitation',
 'TOURS_temp_mean',
 'TOURS_temp_min',
 'TOURS_temp_max',
 'DATE']

In [19]:
# These columns should be deleted from Climates DataFrame
answers.drop(columns=['DATE'], axis=1, inplace =True)
climate.drop(columns=del_colum, axis=1, inplace=True)
answers.shape

(22950, 15)

In [30]:
climate.shape

(22950, 147)

In [32]:
# Two types of observations are missing multiple years for most weather stations.
#Remove 2 types of observations (columns) missing multiple entries for most stations.

col = climate.columns.tolist()
obs = [i.split('_')[-1] for i in col]
obs = pd.Series(obs)
obs.value_counts()

radiation        15
precipitation    15
sunshine         15
mean             15
min              15
max              15
cover            14
humidity         14
pressure         14
speed             9
depth             6
Name: count, dtype: int64

In [34]:
# It seems most stations don't have information about Speed and Depth, So we can drop columns with Speed and Depth in their name
# Searching for columns that have 'Speed' and 'Depth' in their name in climate DataFrame
import re

del_colum2 = []
for i in col:
    if re.search('speed|depth', str(i)):
        del_colum2.append(i)
del_colum2

['BASEL_wind_speed',
 'BASEL_snow_depth',
 'DEBILT_wind_speed',
 'DUSSELDORF_wind_speed',
 'DUSSELDORF_snow_depth',
 'HEATHROW_snow_depth',
 'KASSEL_wind_speed',
 'LJUBLJANA_wind_speed',
 'MAASTRICHT_wind_speed',
 'MADRID_wind_speed',
 'MUNCHENB_snow_depth',
 'OSLO_wind_speed',
 'OSLO_snow_depth',
 'SONNBLICK_wind_speed',
 'VALENTIA_snow_depth']

In [36]:
# These columns should be deleted from Climates DataFrame too
climate.drop(columns=del_colum2, axis=1, inplace=True)
climate.shape

(22950, 132)

In [38]:
# There are three individual observations that need to be filled in. 
# Assume nearby stations have similar weather, then pick one to copy the data from 
# (Ljubljana is near Kassel, Sonnblick is near Munchen, and Olso is close enough to Stockholm).

# Fill in 3 individual observations assuming nearby stations have similar weather.


col = climate.columns.tolist()
complete_stations = []
for i in col:
    if re.search('cover|humidity|pressure', str(i)):
        complete_stations.append(i)
complete_stations = [i.split('_')[0] for i in complete_stations]
complete_stations = pd.Series(complete_stations)
complete_stations.value_counts()

BASEL         3
BELGRADE      3
BUDAPEST      3
DEBILT        3
DUSSELDORF    3
HEATHROW      3
LJUBLJANA     3
MAASTRICHT    3
MADRID        3
OSLO          3
SONNBLICK     3
VALENTIA      3
KASSEL        2
MUNCHENB      2
STOCKHOLM     2
Name: count, dtype: int64

In [40]:
# List all column names from the DataFrame
col = climate.columns.tolist()

# Create a list to hold columns that match specific weather measurements
columns_hold = []
for i in col:
    if re.search('cover|humidity|pressure', str(i)):
        columns_hold.append(i)

# Extract station names and measurements from the matched columns
columns_hold = [[i.split('_')[0], i.split('_')[-1]] for i in columns_hold]

# Find stations with incomplete measurements
uncomplete_measu = []
station_measurement_counts = {}

# Count occurrences of measurements for each station
for station, measurement in columns_hold:
    if station not in station_measurement_counts:
        station_measurement_counts[station] = 0
    station_measurement_counts[station] += 1

# Identify stations with fewer than 3 measurements
for station, count in station_measurement_counts.items():
    if count < 3:
        uncomplete_measu.append(station)

# Output the stations with incomplete observations
uncomplete_measu

['KASSEL', 'MUNCHENB', 'STOCKHOLM']

In [42]:
climate.head()

BASEL_cloud_cover  BASEL_humidity  BASEL_pressure  BASEL_global_radiation  \
0                  7            0.85            1.02                    0.32   
1                  6            0.84            1.02                    0.36   
2                  8            0.90            1.02                    0.18   
3                  3            0.92            1.02                    0.58   
4                  6            0.95            1.02                    0.65   

   BASEL_precipitation  BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  \
0                 0.09            0.70             6.50            0.80   
1                 1.05            1.10             6.10            3.30   
2                 0.30            0.00             8.50            5.10   
3                 0.00            4.10             6.30            3.80   
4                 0.14            5.40             3.00           -0.70   

   BASEL_temp_max  BELGRADE_cloud_cover  BELGRADE_humidity  BELGRADE_pressure  \
0           10.90                     1               0.81               1.02   
1           10.10                     6               0.84               1.02   
2            9.90                     6               0.77               1.02   
3           10.60                     8               0.93               1.03   
4            6.00                     8               0.99               1.03   

   BELGRADE_global_radiation  BELGRADE_precipitation  BELGRADE_sunshine  \
0                       0.88                    0.00               7.00   
1                       0.25                    0.00               0.00   
2                       0.67                    0.00               3.50   
3                       0.25                    0.00               0.00   
4                       0.25                    0.06               0.00   

   BELGRADE_temp_mean  BELGRADE_temp_min  BELGRADE_temp_max  \
0                3.70              -0.90               7.90   
1                2.90               2.20               4.40   
2                3.10              -0.50               6.40   
3                2.00              -2.00               3.00   
4                2.00               0.70               2.80   

   BUDAPEST_cloud_cover  BUDAPEST_humidity  BUDAPEST_pressure  \
0                     4               0.67               1.02   
1                     4               0.67               1.02   
2                     4               0.67               1.02   
3                     4               0.67               1.02   
4                     4               0.67               1.02   

   BUDAPEST_global_radiation  BUDAPEST_precipitation  BUDAPEST_sunshine  \
0                       0.44                    0.01               2.30   
1                       0.18                    0.31               0.00   
2                       0.30                    0.00               0.60   
3                       0.19                    0.00               0.00   
4                       0.19                    0.00               0.00   

   BUDAPEST_temp_mean  BUDAPEST_temp_min  BUDAPEST_temp_max  \
0                2.40              -0.40               5.10   
1                2.30               1.40               3.10   
2                2.70               1.70               5.30   
3                2.00               0.40               4.40   
4                2.50               1.10               5.30   

   DEBILT_cloud_cover  DEBILT_humidity  DEBILT_pressure  \
0                   7             0.85             1.00   
1                   8             0.90             1.01   
2                   6             0.92             1.02   
3                   8             0.95             1.03   
4                   6             0.90             1.02   

   DEBILT_global_radiation  DEBILT_precipitation  DEBILT_sunshine  \
0                     0.07                  0.25             0.00   
1                     0.14              

In [44]:
climate['STOCKHOLM_humidity'] = climate['OSLO_humidity']
climate['MUNCHENB_pressure'] = climate['SONNBLICK_pressure']
climate['KASSEL__cloud_cover'] = climate['LJUBLJANA_cloud_cover']

cols = climate.columns.tolist()
obs = [i.split('_')[-1] for i in cols]
obs = pd.Series(obs)
obs.value_counts()

cover            15
humidity         15
pressure         15
radiation        15
precipitation    15
sunshine         15
mean             15
min              15
max              15
Name: count, dtype: int64

In [46]:
climate.shape

(22950, 135)

In [48]:
answers.shape

(22950, 15)

In [50]:
# Export Cleaned Data Sets
climate.to_csv(os.path.join(path, '02 Data', 'Unsupervised', 'climate_clean.csv'), index=False)
answers.to_csv(os.path.join(path, '02 Data', 'Unsupervised', 'answers_clean.csv'), index=False)

## 3. Reshaping for modeling
- Ensure the layers can be fed to the deep learning model correctly.
- You’ll need to split the observations (X) into 15 groups of 9 types of observations, and your labels (y) should also be in 15 groups (it doesn’t need to be transformed or reshaped).
- The final shapes should be X = (22950, 15, 9) and y = (22950, 15).

In [53]:
X = np.array(climate)
y = np.array(answers)

In [55]:
X = X.reshape(-1,15,9)

In [57]:
X.shape

(22950, 15, 9)

In [59]:
y.shape

(22950, 15)

## 4. Data Split 

In [62]:
# Split data into train and test sets

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [64]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(16065, 15, 9) (16065, 15)
(6885, 15, 9) (6885, 15)


## 5. Creating Keras Model

In [67]:
epochs = 20
batch_size = 32
n_hidden = 64

timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = len(y_train[0])

model = Sequential()
model.add(Conv1D(n_hidden, kernel_size=2, activation='relu', input_shape=(timesteps, input_dim)))
model.add(Dense(16, activation='relu'))
model.add(MaxPooling1D())
model.add(Flatten())
model.add(Dense(n_classes, activation='softmax')) # Options: sigmoid, tanh, softmax, relu

In [69]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d (Conv1D)                 │ (None, 14, 64)         │         1,216 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 14, 16)         │         1,040 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d (MaxPooling1D)    │ (None, 7, 16)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 112)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 15)             │         1,695 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,951 (15.43 KB)

 Trainable params: 3,951 (15.43 KB)

 Non-trainable params: 0 (0.00 B)

## 6. Compiling and Running

In [72]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, batch_size=batch_size, epochs=epochs, verbose=2)

Epoch 1/20
503/503 - 2s - 4ms/step - accuracy: 0.1556 - loss: 2144.9834
Epoch 2/20
503/503 - 1s - 2ms/step - accuracy: 0.1332 - loss: 20923.7852
Epoch 3/20
503/503 - 1s - 2ms/step - accuracy: 0.1272 - loss: 68883.3672
Epoch 4/20
503/503 - 1s - 2ms/step - accuracy: 0.1216 - loss: 144555.8594
Epoch 5/20
503/503 - 1s - 2ms/step - accuracy: 0.1155 - loss: 258281.3438
Epoch 6/20
503/503 - 1s - 2ms/step - accuracy: 0.1139 - loss: 403731.1250
Epoch 7/20
503/503 - 1s - 2ms/step - accuracy: 0.1193 - loss: 608632.3125
Epoch 8/20
503/503 - 1s - 2ms/step - accuracy: 0.1188 - loss: 823492.3750
Epoch 9/20
503/503 - 1s - 2ms/step - accuracy: 0.1217 - loss: 1103585.1250
Epoch 10/20
503/503 - 1s - 2ms/step - accuracy: 0.1182 - loss: 1386911.5000
Epoch 11/20
503/503 - 1s - 2ms/step - accuracy: 0.1194 - loss: 1728296.7500
Epoch 12/20
503/503 - 1s - 2ms/step - accuracy: 0.1134 - loss: 2090528.6250
Epoch 13/20
503/503 - 1s - 2ms/step - accuracy: 0.1162 - loss: 2408040.0000
Epoch 14/20
503/503 - 1s - 2ms/st

In [74]:
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Test Loss: {loss}')
print(f'Test Accuracy: {accuracy}')

216/216 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.0230 - loss: 5820687.5000
Test Loss: 5836553.5
Test Accuracy: 0.02164124883711338


In [76]:
epochs = 40
batch_size = 32
n_hidden = 128

timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = len(y_train[0])

model = Sequential()
model.add(Conv1D(n_hidden, kernel_size=2, activation='relu', input_shape=(timesteps, input_dim)))
model.add(Dense(16, activation='relu'))
model.add(MaxPooling1D())
model.add(Flatten())
model.add(Dense(n_classes, activation='softmax')) # Options: sigmoid, tanh, softmax, relu


model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, batch_size=batch_size, epochs=epochs, verbose=2)


loss, accuracy = model.evaluate(X_test, y_test)
print(f'Test Loss: {loss}')
print(f'Test Accuracy: {accuracy}')

Epoch 1/40
503/503 - 2s - 3ms/step - accuracy: 0.0929 - loss: 3083.4104
Epoch 2/40
503/503 - 1s - 2ms/step - accuracy: 0.1124 - loss: 37806.8633
Epoch 3/40
503/503 - 1s - 2ms/step - accuracy: 0.1075 - loss: 138488.4062
Epoch 4/40
503/503 - 1s - 2ms/step - accuracy: 0.1137 - loss: 294578.1562
Epoch 5/40
503/503 - 1s - 2ms/step - accuracy: 0.1135 - loss: 515793.2188
Epoch 6/40
503/503 - 1s - 2ms/step - accuracy: 0.1127 - loss: 803969.3125
Epoch 7/40
503/503 - 1s - 2ms/step - accuracy: 0.1145 - loss: 1170737.6250
Epoch 8/40
503/503 - 1s - 2ms/step - accuracy: 0.1157 - loss: 1577035.8750
Epoch 9/40
503/503 - 1s - 2ms/step - accuracy: 0.1159 - loss: 2049315.6250
Epoch 10/40
503/503 - 1s - 2ms/step - accuracy: 0.1189 - loss: 2609084.0000
Epoch 11/40
503/503 - 1s - 2ms/step - accuracy: 0.1219 - loss: 3225027.5000
Epoch 12/40
503/503 - 1s - 2ms/step - accuracy: 0.1177 - loss: 3845719.2500
Epoch 13/40
503/503 - 1s - 2ms/step - accuracy: 0.1184 - loss: 4440138.5000
Epoch 14/40
503/503 - 1s - 2ms

In [78]:
epochs = 40
batch_size = 32
n_hidden = 256

timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = len(y_train[0])

model = Sequential()
model.add(Conv1D(n_hidden, kernel_size=2, activation='relu', input_shape=(timesteps, input_dim)))
model.add(Dense(16, activation='relu'))
model.add(MaxPooling1D())
model.add(Flatten())
model.add(Dense(n_classes, activation='softmax')) # Options: sigmoid, tanh, softmax, relu


model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, batch_size=batch_size, epochs=epochs, verbose=2)


loss, accuracy = model.evaluate(X_test, y_test)
print(f'Test Loss: {loss}')
print(f'Test Accuracy: {accuracy}')

Epoch 1/40
503/503 - 2s - 4ms/step - accuracy: 0.0894 - loss: 8452.5107
Epoch 2/40
503/503 - 1s - 2ms/step - accuracy: 0.0880 - loss: 79080.0781
Epoch 3/40
503/503 - 1s - 2ms/step - accuracy: 0.0941 - loss: 281567.4688
Epoch 4/40
503/503 - 1s - 2ms/step - accuracy: 0.1013 - loss: 603424.7500
Epoch 5/40
503/503 - 1s - 2ms/step - accuracy: 0.1018 - loss: 1113079.1250
Epoch 6/40
503/503 - 1s - 2ms/step - accuracy: 0.1025 - loss: 1741979.1250
Epoch 7/40
503/503 - 1s - 2ms/step - accuracy: 0.1020 - loss: 2534203.7500
Epoch 8/40
503/503 - 1s - 2ms/step - accuracy: 0.1010 - loss: 3443804.5000
Epoch 9/40
503/503 - 1s - 2ms/step - accuracy: 0.1018 - loss: 4358989.5000
Epoch 10/40
503/503 - 1s - 2ms/step - accuracy: 0.0985 - loss: 5357783.5000
Epoch 11/40
503/503 - 1s - 2ms/step - accuracy: 0.1016 - loss: 6230577.0000
Epoch 12/40
503/503 - 1s - 2ms/step - accuracy: 0.1054 - loss: 7287021.0000
Epoch 13/40
503/503 - 1s - 2ms/step - accuracy: 0.1063 - loss: 8176967.0000
Epoch 14/40
503/503 - 1s - 2

## 7. Creating Confusion Matrix

In [81]:
# Define list of stations names

stations = {
0: 'BASEL',
1: 'BELGRADE',
2: 'BUDAPEST',
3: 'DEBILT',
4: 'DUSSELDORF',
5: 'HEATHROW',
6: 'KASSEL',
7: 'LJUBLJANA',
8: 'MAASTRICHT',
9: 'MADRID',
10: 'MUNCHENB',
11: 'OSLO',
12: 'SONNBLICK',
13: 'STOCKHOLM',
14: 'VALENTIA'

}

def confusion_matrix(y_true, y_pred):
    y_true = pd.Series([stations[y] for y in np.argmax(y_true, axis=1)])
    y_pred = pd.Series([stations[y] for y in np.argmax(y_pred, axis=1)])

    return pd.crosstab(y_true, y_pred, rownames=['True'], colnames=['Pred'])

In [83]:
# Evaluate

print(confusion_matrix(y_test, model.predict(X_test)))

216/216 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step  
Pred        BASEL  BELGRADE  BUDAPEST  DEBILT  DUSSELDORF  HEATHROW  KASSEL  \
True                                                                          
BASEL          47      3862        40      10         217        44       5   
BELGRADE        0      1305         0       0           0         0       0   
BUDAPEST        0       244         0       0           0         0       0   
DEBILT          0        95         0       0           0         0       0   
DUSSELDORF      0        32         0       0           0         0       0   
HEATHROW        0        98         0       0           0         0       0   
KASSEL          0        13         0       0           0         0       0   
LJUBLJANA       0        71         0       0           0         0       0   
MAASTRICHT      0         9         0       0           0         0       0   
MADRID          1       523         0       0           2         0       0   
MUNCHENB 

# Updated Example Code for LSTM (RNN Approach):

In [86]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
from tensorflow.keras.optimizers import Adam

# Hyperparameters
epochs = 50
batch_size = 64
n_hidden = 128
learning_rate = 0.00005

# Define the LSTM model
model = Sequential()
model.add(LSTM(n_hidden, input_shape=(timesteps, input_dim)))
model.add(Dropout(0.2))
model.add(Dense(n_classes, activation='softmax'))

# Compile with a lower learning rate
optimizer = Adam(learning_rate=learning_rate)
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

# Train the model
model.fit(X_train, y_train, batch_size=batch_size, epochs=epochs, verbose=2)

# Evaluate the model
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Test Loss: {loss}')
print(f'Test Accuracy: {accuracy}')

Epoch 1/50
252/252 - 3s - 13ms/step - accuracy: 0.1309 - loss: 8.9284
Epoch 2/50
252/252 - 2s - 6ms/step - accuracy: 0.1742 - loss: 9.2980
Epoch 3/50
252/252 - 2s - 6ms/step - accuracy: 0.1613 - loss: 9.6242
Epoch 4/50
252/252 - 2s - 6ms/step - accuracy: 0.1445 - loss: 9.6688
Epoch 5/50
252/252 - 2s - 7ms/step - accuracy: 0.1435 - loss: 9.6907
Epoch 6/50
252/252 - 2s - 7ms/step - accuracy: 0.1419 - loss: 9.6878
Epoch 7/50
252/252 - 2s - 7ms/step - accuracy: 0.1371 - loss: 9.6809
Epoch 8/50
252/252 - 2s - 7ms/step - accuracy: 0.1358 - loss: 9.6793
Epoch 9/50
252/252 - 2s - 9ms/step - accuracy: 0.1320 - loss: 9.6617
Epoch 10/50
252/252 - 2s - 7ms/step - accuracy: 0.1244 - loss: 9.6448
Epoch 11/50
252/252 - 2s - 7ms/step - accuracy: 0.1179 - loss: 9.6242
Epoch 12/50
252/252 - 2s - 6ms/step - accuracy: 0.1134 - loss: 9.5911
Epoch 13/50
252/252 - 2s - 6ms/step - accuracy: 0.1143 - loss: 9.5548
Epoch 14/50
252/252 - 2s - 6ms/step - accuracy: 0.1124 - loss: 9.5437
Epoch 15/50
252/252 - 2s - 7

In [88]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout, BatchNormalization
from tensorflow.keras.optimizers import Adam

# Hyperparameters
epochs = 100
batch_size = 64
n_hidden = 128
learning_rate = 0.0001

# Define the LSTM model
model = Sequential()
model.add(LSTM(n_hidden, return_sequences=True, input_shape=(timesteps, input_dim)))
model.add(BatchNormalization())
model.add(LSTM(64))
model.add(Dropout(0.1))
model.add(Dense(n_classes, activation='softmax'))

# Compile with a higher learning rate
optimizer = Adam(learning_rate=learning_rate)
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

# Train the model
model.fit(X_train, y_train, batch_size=batch_size, epochs=epochs, verbose=2)

# Evaluate the model
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Test Loss: {loss}')
print(f'Test Accuracy: {accuracy}')

Epoch 1/100
252/252 - 6s - 23ms/step - accuracy: 0.1293 - loss: 8.9399
Epoch 2/100
252/252 - 3s - 12ms/step - accuracy: 0.1038 - loss: 9.1034
Epoch 3/100
252/252 - 3s - 10ms/step - accuracy: 0.0945 - loss: 9.0778
Epoch 4/100
252/252 - 3s - 10ms/step - accuracy: 0.0954 - loss: 9.0299
Epoch 5/100
252/252 - 3s - 11ms/step - accuracy: 0.0958 - loss: 9.0220
Epoch 6/100
252/252 - 3s - 10ms/step - accuracy: 0.0938 - loss: 8.9764
Epoch 7/100
252/252 - 3s - 11ms/step - accuracy: 0.0982 - loss: 8.9706
Epoch 8/100
252/252 - 3s - 12ms/step - accuracy: 0.0991 - loss: 8.9407
Epoch 9/100
252/252 - 5s - 19ms/step - accuracy: 0.0931 - loss: 8.9417
Epoch 10/100
252/252 - 3s - 13ms/step - accuracy: 0.0968 - loss: 8.9137
Epoch 11/100
252/252 - 3s - 12ms/step - accuracy: 0.0955 - loss: 8.9058
Epoch 12/100
252/252 - 3s - 12ms/step - accuracy: 0.0976 - loss: 8.9056
Epoch 13/100
252/252 - 3s - 11ms/step - accuracy: 0.0930 - loss: 8.9219
Epoch 14/100
252/252 - 3s - 11ms/step - accuracy: 0.0859 - loss: 8.8740
E

In [90]:
print(confusion_matrix(y_test, model.predict(X_test)))

216/216 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step
Pred        BASEL  BELGRADE  BUDAPEST  DEBILT  DUSSELDORF  KASSEL  LJUBLJANA  \
True                                                                           
BASEL           2         2         2       0           3       2          0   
BELGRADE        2         7         1       1           0       0          2   
BUDAPEST        0         1         0       1           0       2          0   
DEBILT          0         0         0       0           0       0          0   
DUSSELDORF      0         0         0       0           0       0          0   
HEATHROW        0         0         1       0           1       0          1   
KASSEL          0         0         0       0           0       0          0   
LJUBLJANA       0         0         0       0           0       0          0   
MAASTRICHT      0         0         0       0           0       0          0   
MADRID          0         0         1       1           0       2          0   

# Scaling Data sets and trying again

In [93]:
from sklearn.preprocessing import StandardScaler

# Scaling the data
scaler = StandardScaler()
scaled_climate = scaler.fit_transform(climate)

# Convert back to DataFrame for easy readability
scaled_climate_df = pd.DataFrame(scaled_climate, columns=climate.columns)

scaled_climate_df.head()

BASEL_cloud_cover  BASEL_humidity  BASEL_pressure  BASEL_global_radiation  \
0               0.66            0.83           -0.00                   -1.10   
1               0.24            0.74           -0.00                   -1.06   
2               1.08            1.28           -0.00                   -1.25   
3              -1.00            1.46           -0.00                   -0.82   
4               0.24            1.73           -0.00                   -0.75   

   BASEL_precipitation  BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  \
0                -0.27           -0.90            -0.53           -0.85   
1                 1.66           -0.81            -0.58           -0.46   
2                 0.16           -1.07            -0.26           -0.19   
3                -0.45           -0.11            -0.56           -0.39   
4                -0.16            0.19            -1.00           -1.08   

   BASEL_temp_max  BELGRADE_cloud_cover  BELGRADE_humidity  BELGRADE_pressure  \
0           -0.48                 -1.21               0.91               0.32   
1           -0.57                  0.65               1.12               0.02   
2           -0.59                  0.65               0.62               0.11   
3           -0.51                  1.40               1.76               1.29   
4           -1.04                  1.40               2.19               1.53   

   BELGRADE_global_radiation  BELGRADE_precipitation  BELGRADE_sunshine  \
0                      -0.74                   -0.37               0.28   
1                      -1.42                   -0.37              -1.33   
2                      -0.97                   -0.37              -0.52   
3                      -1.42                   -0.37              -1.33   
4                      -1.42                   -0.25              -1.33   

   BELGRADE_temp_mean  BELGRADE_temp_min  BELGRADE_temp_max  \
0               -1.02              -1.22              -0.95   
1               -1.11              -0.82              -1.30   
2               -1.08              -1.17              -1.10   
3               -1.21              -1.36              -1.44   
4               -1.21              -1.01              -1.46   

   BUDAPEST_cloud_cover  BUDAPEST_humidity  BUDAPEST_pressure  \
0                 -0.26              -0.01              -0.01   
1                 -0.26              -0.01              -0.01   
2                 -0.26              -0.01              -0.01   
3                 -0.26              -0.01              -0.01   
4                 -0.26              -0.01              -0.01   

   BUDAPEST_global_radiation  BUDAPEST_precipitation  BUDAPEST_sunshine  \
0                      -1.12                   -0.32              -0.77   
1                      -1.42                    0.37              -1.34   
2                      -1.28                   -0.34              -1.20   
3                      -1.40                   -0.34              -1.34   
4                      -1.40                   -0.34              -1.34   

   BUDAPEST_temp_mean  BUDAPEST_temp_min  BUDAPEST_temp_max  \
0               -1.10              -1.12              -1.14   
1               -1.11              -0.88              -1.34   
2               -1.06              -0.84              -1.12   
3               -1.15              -1.01              -1.21   
4               -1.09              -0.92              -1.12   

   DEBILT_cloud_cover  DEBILT_humidity  DEBILT_pressure  \
0                0.71             0.36            -1.23   
1                1.18             0.87            -0.99   
2                0.25             1.07             0.12   
3                1.18             1.38             1.14   
4                0.25             0.87             0.92   

   DEBILT_global_radiation  DEBILT_precipitation  DEBILT_sunshine  \
0                    -1.20                  0.05            -1.08   
1                    -1.12              

In [95]:
X = np.array(scaled_climate_df)
y = np.array(answers)

In [97]:
X = X.reshape(-1,15,9)

In [99]:
X.shape

(22950, 15, 9)

In [101]:
y.shape

(22950, 15)

In [103]:
# Split data into train and test sets

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [105]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(16065, 15, 9) (16065, 15)
(6885, 15, 9) (6885, 15)


In [107]:

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv1D, Dense, Dropout, MaxPooling1D, Flatten, BatchNormalization
from tensorflow.keras.optimizers import Adam

# Hyperparameters
epochs = 50
batch_size = 64
n_hidden = 64
learning_rate = 0.00005

# Define the model
model = Sequential()
model.add(Conv1D(n_hidden, kernel_size=2, activation='relu', input_shape=(timesteps, input_dim)))
model.add(BatchNormalization())
model.add(Dropout(0.3))
model.add(Dense(16, activation='relu'))
model.add(MaxPooling1D())
model.add(Flatten())
model.add(Dense(n_classes, activation='softmax'))

# Compile with a lower learning rate
optimizer = Adam(learning_rate=learning_rate)
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

# Train the model
model.fit(X_train, y_train, batch_size=batch_size, epochs=epochs, verbose=2)

# Evaluate the model
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Test Loss: {loss}')
print(f'Test Accuracy: {accuracy}')

Epoch 1/50
252/252 - 2s - 7ms/step - accuracy: 0.0942 - loss: 11.7099
Epoch 2/50
252/252 - 1s - 3ms/step - accuracy: 0.1030 - loss: 12.4624
Epoch 3/50
252/252 - 1s - 3ms/step - accuracy: 0.1174 - loss: 14.7660
Epoch 4/50
252/252 - 1s - 3ms/step - accuracy: 0.1254 - loss: 18.0082
Epoch 5/50
252/252 - 1s - 3ms/step - accuracy: 0.1318 - loss: 21.6980
Epoch 6/50
252/252 - 1s - 3ms/step - accuracy: 0.1327 - loss: 26.1547
Epoch 7/50
252/252 - 1s - 3ms/step - accuracy: 0.1226 - loss: 31.1660
Epoch 8/50
252/252 - 1s - 2ms/step - accuracy: 0.1163 - loss: 36.3145
Epoch 9/50
252/252 - 1s - 3ms/step - accuracy: 0.1087 - loss: 42.1711
Epoch 10/50
252/252 - 1s - 3ms/step - accuracy: 0.1079 - loss: 48.4663
Epoch 11/50
252/252 - 1s - 2ms/step - accuracy: 0.1066 - loss: 56.3213
Epoch 12/50
252/252 - 1s - 2ms/step - accuracy: 0.1039 - loss: 65.1427
Epoch 13/50
252/252 - 1s - 3ms/step - accuracy: 0.1011 - loss: 75.6762
Epoch 14/50
252/252 - 1s - 2ms/step - accuracy: 0.0998 - loss: 86.0898
Epoch 15/50
252

In [109]:
# Evaluate

print(confusion_matrix(y_test, model.predict(X_test)))

216/216 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
Pred        BASEL  BELGRADE  BUDAPEST  DEBILT  DUSSELDORF  HEATHROW  KASSEL  \
True                                                                          
BASEL         448       490       396     507         115       130      63   
BELGRADE       70       219       116      53          47        89      19   
BUDAPEST        9        35        32       2           6        11       0   
DEBILT          6        11        18       0           2         3       0   
DUSSELDORF      1         7         5       2           1         0       0   
HEATHROW        4        15         7       8           1         2       1   
KASSEL          0         2         2       0           1         1       0   
LJUBLJANA       3        14         5       5           0         3       3   
MAASTRICHT      0         1         3       3           0         0       0   
MADRID         28        86        86      33           6         7      31   
MUNCHENB   

In [111]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout, BatchNormalization, Bidirectional
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import ReduceLROnPlateau

# Hyperparameters
epochs = 100
batch_size = 64
n_hidden = 64
learning_rate = 0.0001

# Define the LSTM model
model = Sequential()
model.add(Bidirectional(LSTM(n_hidden, return_sequences=True, input_shape=(timesteps, input_dim))))
model.add(BatchNormalization())
model.add(Dropout(0.3))
model.add(LSTM(32))
model.add(Dense(n_classes, activation='softmax'))

# Compile with a lower learning rate
optimizer = Adam(learning_rate=learning_rate)
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

# Learning rate scheduler to reduce the learning rate if loss does not improve
reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=5, min_lr=1e-6)

# Train the model
model.fit(X_train, y_train, batch_size=batch_size, epochs=epochs, verbose=2, callbacks=[reduce_lr], validation_data=(X_test, y_test))

# Evaluate the model
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Test Loss: {loss}')
print(f'Test Accuracy: {accuracy}')

Epoch 1/100
252/252 - 8s - 31ms/step - accuracy: 0.0960 - loss: 8.9117 - val_accuracy: 0.0606 - val_loss: 8.9382 - learning_rate: 1.0000e-04
Epoch 2/100
252/252 - 3s - 12ms/step - accuracy: 0.0649 - loss: 9.1750 - val_accuracy: 0.0671 - val_loss: 8.9660 - learning_rate: 1.0000e-04
Epoch 3/100
252/252 - 3s - 13ms/step - accuracy: 0.0609 - loss: 9.1164 - val_accuracy: 0.0558 - val_loss: 8.9701 - learning_rate: 1.0000e-04
Epoch 4/100
252/252 - 2s - 10ms/step - accuracy: 0.0515 - loss: 9.1235 - val_accuracy: 0.0516 - val_loss: 8.9644 - learning_rate: 1.0000e-04
Epoch 5/100
252/252 - 2s - 10ms/step - accuracy: 0.0489 - loss: 9.1031 - val_accuracy: 0.0513 - val_loss: 8.9445 - learning_rate: 1.0000e-04
Epoch 6/100
252/252 - 3s - 11ms/step - accuracy: 0.0484 - loss: 9.0821 - val_accuracy: 0.0505 - val_loss: 8.9344 - learning_rate: 1.0000e-04
Epoch 7/100
252/252 - 3s - 10ms/step - accuracy: 0.0481 - loss: 9.0710 - val_accuracy: 0.0510 - val_loss: 8.9168 - learning_rate: 1.0000e-04
Epoch 8/100
2

In [113]:
# Evaluate

print(confusion_matrix(y_test, model.predict(X_test)))

216/216 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step
Pred        BELGRADE  BUDAPEST  DUSSELDORF  KASSEL  LJUBLJANA  MADRID  \
True                                                                    
BASEL              5         1           1       1          0    1764   
BELGRADE           7         4           0       1          0    1184   
BUDAPEST           1         1           0       0          0     231   
DEBILT             0         0           0       0          0      96   
DUSSELDORF         1         0           0       0          0      30   
HEATHROW           0         0           0       0          0      84   
KASSEL             0         0           0       0          0      13   
LJUBLJANA          0         0           0       1          0      61   
MAASTRICHT         0         0           0       0          0       3   
MADRID             0         0           0       1          2     235   
MUNCHENB           1         0           0       0          0       4   
OSLO      